{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# UPEC population structure and antimicrobial susceptibility in Norfolk, UK"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Globals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import glob\n",
    "import re\n",
    "from os import listdir, path, rename, system\n",
    "import matplotlib.pyplot as plt \n",
    "import numpy as np \n",
    "import seaborn as sns\n",
    "\n",
    "from scipy.stats import hypergeom, fisher_exact, chi2_contingency\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "\n",
    "def rename_isolate(i):\n",
    "    if i == \"Reference\":\n",
    "        return i\n",
    "    try:\n",
    "        return re.search(\"GL\\d+\", i).group(0)\n",
    "    except AttributeError:\n",
    "        *_, GL = i.split(\"-\")\n",
    "        GL = \"GL\" + GL\n",
    "        return rename_isolate(GL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = r'C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\manuscripts\\UPEC Population Study\\Table S1.xlsx'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### In silico typing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# phylogrouping\n",
    "\n",
    "outdir = 'dataset/phylogroup/'\n",
    "\n",
    "def EzClerment(fp, output, script=\"EzClermont.py\"):\n",
    "    OUT = path.basename(fp).replace(\".fasta\", \".txt\")\n",
    "    system(f'python \"{script}\" \"{fp}\" 1> \"{output}{OUT}\" 2>&1')\n",
    "\n",
    "\n",
    "def run_phylo(fp, out, script=\"EzClermont.py\"):\n",
    "    for f in glob.glob(path.join(fp, \"*.fasta\")):\n",
    "        EzClerment(fp=f, output=out, script=script)\n",
    "\n",
    "run_phylo(\"dataset/FASTA/\", \"dataset/phylogroup/\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phylogroups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(B2    152\n",
       " D      28\n",
       " B1     19\n",
       " A       6\n",
       " F       6\n",
       " G       2\n",
       " C       2\n",
       " E       2\n",
       " Name: phylogroup, dtype: int64,\n",
       " B2    70.0\n",
       " D     12.9\n",
       " B1     8.8\n",
       " A      2.8\n",
       " F      2.8\n",
       " G      0.9\n",
       " C      0.9\n",
       " E      0.9\n",
       " Name: phylogroup, dtype: float64)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylogroups = pd.read_excel(r\"C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\manuscripts\\UPEC Population Study\\Table S1.xlsx\")['phylogroup']\n",
    "phylogroups.value_counts(),phylogroups.value_counts(normalize=True).mul(100).round(1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequence types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "st = pd.read_excel(r\"C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\manuscripts\\UPEC Population Study\\Table S1.xlsx\")[['ID','ST']]\n",
    "st = st.fillna(0)\n",
    "st['ST'] = st['ST'].astype('int32')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "73      28\n",
       "12      20\n",
       "69      19\n",
       "131     18\n",
       "404     14\n",
       "        ..\n",
       "372      1\n",
       "108      1\n",
       "3177     1\n",
       "357      1\n",
       "104      1\n",
       "Name: ST, Length: 75, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "st['ST'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "57.14285714285714"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "st['ST'].value_counts()[:8].sum() / 217 * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(73      28\n",
       " 12      20\n",
       " 69      19\n",
       " 131     18\n",
       " 404     14\n",
       " 95      12\n",
       " 127      7\n",
       " 1193     6\n",
       " Name: ST, dtype: int64,\n",
       " 73      12.9\n",
       " 12       9.2\n",
       " 69       8.8\n",
       " 131      8.3\n",
       " 404      6.5\n",
       " 95       5.5\n",
       " 127      3.2\n",
       " 1193     2.8\n",
       " Name: ST, dtype: float64)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "st['ST'].value_counts()[:8], st['ST'].value_counts(normalize=True).mul(100).round(1)[:8]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Core genome alignment summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# core alignment summary\n",
    "ca = pd.read_csv('dataset\\core_alignment_summary.txt', delimiter='\\t')\n",
    "ca = ca[:-1]\n",
    "(ca['ALIGNED'] / ca['LENGTH']).describe()\n",
    "# count      217\n",
    "# mean       0.843553\n",
    "# std        0.055279\n",
    "# min        0.728436\n",
    "# 25%        0.782434\n",
    "# 50%        0.858234\n",
    "# 75%        0.890663\n",
    "# max        0.941170"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AMR rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "amr = pd.read_excel(r\"C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\manuscripts\\UPEC Population Study\\Table S1.xlsx\", usecols=\"A,E,I,J,K,L,M,N,O\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "amr = amr.iloc[18:,1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ABX</th>\n",
       "      <th>Overall</th>\n",
       "      <th>Community</th>\n",
       "      <th>Hospital</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMP</td>\n",
       "      <td>39.2</td>\n",
       "      <td>27.6</td>\n",
       "      <td>52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AUG</td>\n",
       "      <td>7.6</td>\n",
       "      <td>4.8</td>\n",
       "      <td>10.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CLX</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.9</td>\n",
       "      <td>10.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CIP</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.6</td>\n",
       "      <td>10.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GEN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NIT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TRI</td>\n",
       "      <td>25.1</td>\n",
       "      <td>15.2</td>\n",
       "      <td>36.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ABX  Overall  Community  Hospital\n",
       "0  AMP     39.2       27.6      52.1\n",
       "1  AUG      7.6        4.8      10.9\n",
       "2  CLX      6.0        1.9      10.6\n",
       "3  CIP      9.0        7.6      10.6\n",
       "4  GEN      7.0        1.0      13.8\n",
       "5  NIT      1.0        0.0       2.1\n",
       "6  TRI     25.1       15.2      36.2"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rx_rates = {\n",
    "    \"ABX\" : [],\n",
    "    \"Overall\" : [],\n",
    "    \"Community\" : [],\n",
    "    \"Hospital\" : []\n",
    "    }\n",
    "\n",
    "for col in amr.columns[1:]:\n",
    "    rx_rates['ABX'].append(col)\n",
    "    rx_rates['Overall'].append(amr.loc[:,col].value_counts(normalize=True)['R'].round(3) * 100)\n",
    "    rx_rates['Hospital'].append(amr.loc[amr[\"Source\"]=='Hospital',col].value_counts(normalize=True)['R'].round(3) * 100)\n",
    "    try:\n",
    "        rx_rates['Community'].append(amr.loc[amr[\"Source\"]=='Community',col].value_counts(normalize=True)['R'].round(3) * 100)\n",
    "    except KeyError:\n",
    "        rx_rates['Community'].append(0)\n",
    "\n",
    "pd.DataFrame(rx_rates)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Relationship between antibiotic resistance and source"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pheno</th>\n",
       "      <th>res_com</th>\n",
       "      <th>res_hosp</th>\n",
       "      <th>sens_com</th>\n",
       "      <th>sens_hosp</th>\n",
       "      <th>chi2</th>\n",
       "      <th>p_chi</th>\n",
       "      <th>dof</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMP</td>\n",
       "      <td>29</td>\n",
       "      <td>49</td>\n",
       "      <td>76</td>\n",
       "      <td>45</td>\n",
       "      <td>12.500509</td>\n",
       "      <td>0.000407</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AUG</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>100</td>\n",
       "      <td>82</td>\n",
       "      <td>2.600342</td>\n",
       "      <td>0.106841</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GEN</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>104</td>\n",
       "      <td>81</td>\n",
       "      <td>12.575558</td>\n",
       "      <td>0.000391</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NIT</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>105</td>\n",
       "      <td>92</td>\n",
       "      <td>2.256723</td>\n",
       "      <td>0.133035</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRI</td>\n",
       "      <td>16</td>\n",
       "      <td>34</td>\n",
       "      <td>89</td>\n",
       "      <td>60</td>\n",
       "      <td>11.551551</td>\n",
       "      <td>0.000677</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CLX</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>103</td>\n",
       "      <td>84</td>\n",
       "      <td>6.676173</td>\n",
       "      <td>0.009771</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CIP</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>97</td>\n",
       "      <td>84</td>\n",
       "      <td>0.549563</td>\n",
       "      <td>0.458496</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Pheno  res_com  res_hosp  sens_com  sens_hosp       chi2     p_chi  dof\n",
       "0   AMP       29        49        76         45  12.500509  0.000407    1\n",
       "1   AUG        5        10       100         82   2.600342  0.106841    1\n",
       "2   GEN        1        13       104         81  12.575558  0.000391    1\n",
       "3   NIT        0         2       105         92   2.256723  0.133035    1\n",
       "4   TRI       16        34        89         60  11.551551  0.000677    1\n",
       "5   CLX        2        10       103         84   6.676173  0.009771    1\n",
       "6   CIP        8        10        97         84   0.549563  0.458496    1"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def build_matrix_and_text(pheno):\n",
    "    res_com = len(amr[(amr[pheno] == 'R') & (amr['Source']=='Community')])\n",
    "    res_hosp = len(amr[(amr[pheno]== 'R') & (amr['Source']=='Hospital')])\n",
    "    sens_com = len(amr[(amr[pheno] == 'S') & (amr['Source']=='Community')])\n",
    "    sens_hosp = len(amr[(amr[pheno] == 'S') & (amr['Source']=='Hospital')])\n",
    "    table = np.array([[res_com, res_hosp], [sens_com, sens_hosp]])\n",
    "    chi2, p_chi, dof, _ = chi2_contingency(table, correction=False)\n",
    "    return res_com, res_hosp, sens_com, sens_hosp, chi2, p_chi, dof\n",
    "\n",
    "data = []\n",
    "for r in ['AMP', 'AUG',\t'GEN', 'NIT', 'TRI', 'CLX', 'CIP']:\n",
    "    result = build_matrix_and_text(r)\n",
    "    data.append([r]+list(result))\n",
    "\n",
    "pd.DataFrame(data, columns=['Pheno', \"res_com\", 'res_hosp', 'sens_com', 'sens_hosp', 'chi2', 'p_chi', 'dof'])\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AMR determinants associated with phenotypic resistance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pheno = pd.read_csv(r\"C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\Data\\UPEC population study\\gene_name_assoc.csv\", index_col=0)\n",
    "gene_pres_abs = pd.read_excel(r\"C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\manuscripts\\UPEC Population Study\\code\\AMR_genes_pres_abs.xlsx\", index_col=0)\n",
    "amr = pd.read_excel(r\"C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\manuscripts\\UPEC Population Study\\Table S1.xlsx\", usecols=\"A,I,J,K,L,M,N,O\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pres_abs = gene_pres_abs.append(pd.Series(name='GL234', data=[0]*81, index=gene_pres_abs.columns, dtype=int))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno_genes = {pheno : [] for pheno in amr.columns[1:]}\n",
    "for gene, pheno1, pheno2, pheno3 in gene_pheno.itertuples():\n",
    "    pheno_genes[pheno1].append(gene)\n",
    "    if not isinstance(pheno2, float):\n",
    "        pheno_genes[pheno2].append(gene)\n",
    "    if not isinstance(pheno3, float):\n",
    "        pheno_genes[pheno3].append(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['GL1', 'GL2', 'GL3', 'GL4', 'GL5', 'GL6', 'GL7', 'GL8', 'GL9', 'GL10',\n",
       "       ...\n",
       "       'GL236', 'GL237', 'GL238', 'GL239', 'GL240', 'GL243', 'GL244', 'GL245',\n",
       "       'GL246', 'GL247'],\n",
       "      dtype='object', name='ID', length=217)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amr.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['GL70', 'GL87', 'GL127', 'GL146', 'GL153', 'GL163', 'GL164', 'GL183',\n",
       "       'GL204', 'GL239', 'GL240'],\n",
       "      dtype='object', name='Name')"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pres_abs[gene_pres_abs['blaCTX-M-15']==1].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_matrix_and_text(pheno, geno):\n",
    "    res_gene = len(amr[(amr[pheno] == 'R') & (amr.index.isin(gene_pres_abs[gene_pres_abs[geno]==1].index))])\n",
    "    res_no_gene = len(amr[(amr[pheno]== 'R') & (~amr.index.isin(gene_pres_abs[gene_pres_abs[geno]==1].index))])\n",
    "    sens_gene = len(amr[(amr[pheno] == 'S') & (amr.index.isin(gene_pres_abs[gene_pres_abs[geno]==1].index))])\n",
    "    sens_no_gene = len(amr[(amr[pheno] == 'S') & (~amr.index.isin(gene_pres_abs[gene_pres_abs[geno]==1].index))])\n",
    "    table = np.array([[res_gene, res_no_gene], [sens_gene, sens_no_gene]])\n",
    "    oddsr, p_fisher = fisher_exact(table, alternative='two-sided')\n",
    "    chi2, p_chi, dof, _ = chi2_contingency(table, correction=False)\n",
    "    return res_gene, res_no_gene, sens_gene, sens_no_gene, oddsr, p_fisher, chi2, p_chi, dof"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cramer_assoc(p):\n",
    "    phi2 = p / 217\n",
    "    cramer_assoc = np.sqrt(phi2)\n",
    "    return cramer_assoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pheno_assoc = {\n",
    "    \"gene\" : [],\n",
    "    \"pheno\" : [],\n",
    "    \"gene_pres_pheno_pos\" : [],\n",
    "    \"gene_abs_pheno_pos\" : [],\n",
    "    \"gene_pres_pheno_neg\" : [],\n",
    "    \"gene_abs_pheno_neg\" : [],\n",
    "    \"oddsr\" : [],\n",
    "    \"p-value\" : [],\n",
    "    \"chi2\" : [],\n",
    "    \"p_chi\" : [],\n",
    "    \"dof\" : []\n",
    "    }\n",
    "for pheno,genes in pheno_genes.items():\n",
    "    for gene in genes:\n",
    "        result = build_matrix_and_text(pheno, gene)\n",
    "        gene_pheno_assoc[\"gene\"].append(gene)\n",
    "        gene_pheno_assoc[\"pheno\"].append(pheno)\n",
    "        for key, val in zip(list(gene_pheno_assoc.keys())[2:], result):\n",
    "            gene_pheno_assoc[key].append(val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "fet = pd.DataFrame(gene_pheno_assoc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "fet['cramer_assoc'] = fet['chi2'].apply(get_cramer_assoc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_value_corr = fdrcorrection(fet['p-value'])\n",
    "fet = fet.assign(H0=p_value_corr[0], p_corrected=p_value_corr[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_value_corr = fdrcorrection(fet['p_chi'])\n",
    "fet = fet.assign(H0_chi=p_value_corr[0], p_chi_corrected=p_value_corr[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pheno</th>\n",
       "      <th>gene_pres_pheno_pos</th>\n",
       "      <th>gene_abs_pheno_pos</th>\n",
       "      <th>gene_pres_pheno_neg</th>\n",
       "      <th>gene_abs_pheno_neg</th>\n",
       "      <th>oddsr</th>\n",
       "      <th>p-value</th>\n",
       "      <th>chi2</th>\n",
       "      <th>p_chi</th>\n",
       "      <th>dof</th>\n",
       "      <th>cramer_assoc</th>\n",
       "      <th>H0</th>\n",
       "      <th>p_corrected</th>\n",
       "      <th>H0_chi</th>\n",
       "      <th>p_chi_corrected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>dfrA1</td>\n",
       "      <td>TRI</td>\n",
       "      <td>22</td>\n",
       "      <td>37</td>\n",
       "      <td>3</td>\n",
       "      <td>155</td>\n",
       "      <td>30.720721</td>\n",
       "      <td>1.510279e-11</td>\n",
       "      <td>52.780564</td>\n",
       "      <td>3.729738e-13</td>\n",
       "      <td>1</td>\n",
       "      <td>0.493182</td>\n",
       "      <td>True</td>\n",
       "      <td>1.711649e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>1.950940e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>dfrA17</td>\n",
       "      <td>TRI</td>\n",
       "      <td>14</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>158</td>\n",
       "      <td>inf</td>\n",
       "      <td>3.465753e-09</td>\n",
       "      <td>40.077148</td>\n",
       "      <td>2.441282e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.429753</td>\n",
       "      <td>True</td>\n",
       "      <td>3.366731e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>1.106715e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>dfrA12</td>\n",
       "      <td>TRI</td>\n",
       "      <td>7</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>158</td>\n",
       "      <td>inf</td>\n",
       "      <td>8.367805e-05</td>\n",
       "      <td>19.370621</td>\n",
       "      <td>1.076505e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>0.298773</td>\n",
       "      <td>True</td>\n",
       "      <td>3.556317e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>3.852753e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>dfrA5</td>\n",
       "      <td>TRI</td>\n",
       "      <td>5</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>158</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.307857e-03</td>\n",
       "      <td>13.705628</td>\n",
       "      <td>2.138127e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>0.251316</td>\n",
       "      <td>True</td>\n",
       "      <td>4.680751e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>6.321420e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>dfrA14</td>\n",
       "      <td>TRI</td>\n",
       "      <td>6</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>157</td>\n",
       "      <td>17.773585</td>\n",
       "      <td>1.829865e-03</td>\n",
       "      <td>12.515172</td>\n",
       "      <td>4.036606e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>0.240153</td>\n",
       "      <td>True</td>\n",
       "      <td>5.925278e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>1.055728e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>aac(3)-IIe</td>\n",
       "      <td>GEN</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>202</td>\n",
       "      <td>inf</td>\n",
       "      <td>3.700457e-08</td>\n",
       "      <td>83.097630</td>\n",
       "      <td>7.809926e-20</td>\n",
       "      <td>1</td>\n",
       "      <td>0.618820</td>\n",
       "      <td>True</td>\n",
       "      <td>2.516311e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>7.586785e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>aac(3)-IId</td>\n",
       "      <td>GEN</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>199</td>\n",
       "      <td>58.041667</td>\n",
       "      <td>1.710256e-07</td>\n",
       "      <td>64.841590</td>\n",
       "      <td>8.116776e-16</td>\n",
       "      <td>1</td>\n",
       "      <td>0.546634</td>\n",
       "      <td>True</td>\n",
       "      <td>8.945953e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>5.017643e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>aac(6')-Ib-cr5</td>\n",
       "      <td>GEN</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>201</td>\n",
       "      <td>100.500000</td>\n",
       "      <td>4.521959e-06</td>\n",
       "      <td>56.005356</td>\n",
       "      <td>7.227386e-14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.508025</td>\n",
       "      <td>True</td>\n",
       "      <td>2.196380e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>4.095519e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>aadA5</td>\n",
       "      <td>GEN</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>26.266667</td>\n",
       "      <td>1.417312e-05</td>\n",
       "      <td>40.858096</td>\n",
       "      <td>1.636937e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.433920</td>\n",
       "      <td>True</td>\n",
       "      <td>6.425147e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>7.950836e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>aadA2</td>\n",
       "      <td>GEN</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>197</td>\n",
       "      <td>14.327273</td>\n",
       "      <td>1.548610e-03</td>\n",
       "      <td>20.555082</td>\n",
       "      <td>5.793977e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>0.307772</td>\n",
       "      <td>True</td>\n",
       "      <td>5.265273e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>2.188836e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>blaCTX-M-15</td>\n",
       "      <td>CLX</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>204</td>\n",
       "      <td>1020.000000</td>\n",
       "      <td>1.392117e-14</td>\n",
       "      <td>161.686247</td>\n",
       "      <td>4.844348e-37</td>\n",
       "      <td>1</td>\n",
       "      <td>0.863191</td>\n",
       "      <td>True</td>\n",
       "      <td>1.899112e-13</td>\n",
       "      <td>True</td>\n",
       "      <td>1.647078e-35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>blaOXA-1</td>\n",
       "      <td>CLX</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>205</td>\n",
       "      <td>inf</td>\n",
       "      <td>6.831613e-09</td>\n",
       "      <td>105.414692</td>\n",
       "      <td>9.907259e-25</td>\n",
       "      <td>1</td>\n",
       "      <td>0.696981</td>\n",
       "      <td>True</td>\n",
       "      <td>5.806871e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>1.684234e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>gyrA_D87N</td>\n",
       "      <td>CIP</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>199</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.571374e-20</td>\n",
       "      <td>178.147690</td>\n",
       "      <td>1.229890e-40</td>\n",
       "      <td>1</td>\n",
       "      <td>0.906067</td>\n",
       "      <td>True</td>\n",
       "      <td>5.342672e-19</td>\n",
       "      <td>True</td>\n",
       "      <td>8.363249e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>parC_S80I</td>\n",
       "      <td>CIP</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>197</td>\n",
       "      <td>492.500000</td>\n",
       "      <td>2.077837e-18</td>\n",
       "      <td>154.954710</td>\n",
       "      <td>1.432488e-35</td>\n",
       "      <td>1</td>\n",
       "      <td>0.845031</td>\n",
       "      <td>True</td>\n",
       "      <td>4.709765e-17</td>\n",
       "      <td>True</td>\n",
       "      <td>3.246974e-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>gyrA_S83L</td>\n",
       "      <td>CIP</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>186</td>\n",
       "      <td>114.461538</td>\n",
       "      <td>1.396406e-14</td>\n",
       "      <td>96.699079</td>\n",
       "      <td>8.070680e-23</td>\n",
       "      <td>1</td>\n",
       "      <td>0.667546</td>\n",
       "      <td>True</td>\n",
       "      <td>1.899112e-13</td>\n",
       "      <td>True</td>\n",
       "      <td>9.146770e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>parC_E84V</td>\n",
       "      <td>CIP</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>199</td>\n",
       "      <td>inf</td>\n",
       "      <td>7.805876e-09</td>\n",
       "      <td>79.968519</td>\n",
       "      <td>3.804229e-19</td>\n",
       "      <td>1</td>\n",
       "      <td>0.607057</td>\n",
       "      <td>True</td>\n",
       "      <td>5.897773e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>3.233595e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>aac(6')-Ib-cr5</td>\n",
       "      <td>CIP</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>199</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.372533e-07</td>\n",
       "      <td>68.219589</td>\n",
       "      <td>1.462655e-16</td>\n",
       "      <td>1</td>\n",
       "      <td>0.560692</td>\n",
       "      <td>True</td>\n",
       "      <td>7.777688e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>9.946051e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>parE_L416F</td>\n",
       "      <td>CIP</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>199</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.372533e-07</td>\n",
       "      <td>68.219589</td>\n",
       "      <td>1.462655e-16</td>\n",
       "      <td>1</td>\n",
       "      <td>0.560692</td>\n",
       "      <td>True</td>\n",
       "      <td>7.777688e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>9.946051e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>parE_I529L</td>\n",
       "      <td>CIP</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>188</td>\n",
       "      <td>10.876033</td>\n",
       "      <td>1.469030e-04</td>\n",
       "      <td>24.151515</td>\n",
       "      <td>8.904608e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>0.333613</td>\n",
       "      <td>True</td>\n",
       "      <td>5.876122e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>3.784458e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>blaTEM-40</td>\n",
       "      <td>AUG</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>197</td>\n",
       "      <td>inf</td>\n",
       "      <td>6.650728e-03</td>\n",
       "      <td>22.094418</td>\n",
       "      <td>2.595634e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>0.319089</td>\n",
       "      <td>True</td>\n",
       "      <td>1.966302e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>1.038253e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>blaTEM-1</td>\n",
       "      <td>AMP</td>\n",
       "      <td>55</td>\n",
       "      <td>30</td>\n",
       "      <td>4</td>\n",
       "      <td>128</td>\n",
       "      <td>58.666667</td>\n",
       "      <td>1.214743e-24</td>\n",
       "      <td>99.350479</td>\n",
       "      <td>2.115466e-23</td>\n",
       "      <td>1</td>\n",
       "      <td>0.676636</td>\n",
       "      <td>True</td>\n",
       "      <td>8.260254e-23</td>\n",
       "      <td>True</td>\n",
       "      <td>2.877034e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>blaCTX-M-15</td>\n",
       "      <td>AMP</td>\n",
       "      <td>10</td>\n",
       "      <td>75</td>\n",
       "      <td>1</td>\n",
       "      <td>131</td>\n",
       "      <td>17.466667</td>\n",
       "      <td>4.465471e-04</td>\n",
       "      <td>13.017899</td>\n",
       "      <td>3.085278e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>0.244929</td>\n",
       "      <td>True</td>\n",
       "      <td>1.686956e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>8.391957e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>blaOXA-1</td>\n",
       "      <td>AMP</td>\n",
       "      <td>6</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>132</td>\n",
       "      <td>inf</td>\n",
       "      <td>3.233583e-03</td>\n",
       "      <td>9.582604</td>\n",
       "      <td>1.964296e-03</td>\n",
       "      <td>1</td>\n",
       "      <td>0.210142</td>\n",
       "      <td>True</td>\n",
       "      <td>9.994710e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>4.308779e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>blaSHV-1</td>\n",
       "      <td>AMP</td>\n",
       "      <td>5</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>132</td>\n",
       "      <td>inf</td>\n",
       "      <td>8.568994e-03</td>\n",
       "      <td>7.947836</td>\n",
       "      <td>4.814492e-03</td>\n",
       "      <td>1</td>\n",
       "      <td>0.191379</td>\n",
       "      <td>True</td>\n",
       "      <td>2.427882e-02</td>\n",
       "      <td>True</td>\n",
       "      <td>1.023080e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene pheno  gene_pres_pheno_pos  gene_abs_pheno_pos  \\\n",
       "59           dfrA1   TRI                   22                  37   \n",
       "61          dfrA17   TRI                   14                  45   \n",
       "63          dfrA12   TRI                    7                  52   \n",
       "64           dfrA5   TRI                    5                  54   \n",
       "65          dfrA14   TRI                    6                  53   \n",
       "55      aac(3)-IIe   GEN                    6                   9   \n",
       "51      aac(3)-IId   GEN                    7                   8   \n",
       "53  aac(6')-Ib-cr5   GEN                    5                  10   \n",
       "48           aadA5   GEN                    6                   9   \n",
       "52           aadA2   GEN                    4                  11   \n",
       "25     blaCTX-M-15   CLX                   10                   2   \n",
       "26        blaOXA-1   CLX                    6                   6   \n",
       "33       gyrA_D87N   CIP                   15                   3   \n",
       "31       parC_S80I   CIP                   15                   3   \n",
       "32       gyrA_S83L   CIP                   16                   2   \n",
       "37       parC_E84V   CIP                    7                  11   \n",
       "38  aac(6')-Ib-cr5   CIP                    6                  12   \n",
       "40      parE_L416F   CIP                    6                  12   \n",
       "35      parE_I529L   CIP                    7                  11   \n",
       "18       blaTEM-40   AUG                    2                  16   \n",
       "1         blaTEM-1   AMP                   55                  30   \n",
       "3      blaCTX-M-15   AMP                   10                  75   \n",
       "6         blaOXA-1   AMP                    6                  79   \n",
       "2         blaSHV-1   AMP                    5                  80   \n",
       "\n",
       "    gene_pres_pheno_neg  gene_abs_pheno_neg        oddsr       p-value  \\\n",
       "59                    3                 155    30.720721  1.510279e-11   \n",
       "61                    0                 158          inf  3.465753e-09   \n",
       "63                    0                 158          inf  8.367805e-05   \n",
       "64                    0                 158          inf  1.307857e-03   \n",
       "65                    1                 157    17.773585  1.829865e-03   \n",
       "55                    0                 202          inf  3.700457e-08   \n",
       "51                    3                 199    58.041667  1.710256e-07   \n",
       "53                    1                 201   100.500000  4.521959e-06   \n",
       "48                    5                 197    26.266667  1.417312e-05   \n",
       "52                    5                 197    14.327273  1.548610e-03   \n",
       "25                    1                 204  1020.000000  1.392117e-14   \n",
       "26                    0                 205          inf  6.831613e-09   \n",
       "33                    0                 199          inf  1.571374e-20   \n",
       "31                    2                 197   492.500000  2.077837e-18   \n",
       "32                   13                 186   114.461538  1.396406e-14   \n",
       "37                    0                 199          inf  7.805876e-09   \n",
       "38                    0                 199          inf  1.372533e-07   \n",
       "40                    0                 199          inf  1.372533e-07   \n",
       "35                   11                 188    10.876033  1.469030e-04   \n",
       "18                    0                 197          inf  6.650728e-03   \n",
       "1                     4                 128    58.666667  1.214743e-24   \n",
       "3                     1                 131    17.466667  4.465471e-04   \n",
       "6                     0                 132          inf  3.233583e-03   \n",
       "2                     0                 132          inf  8.568994e-03   \n",
       "\n",
       "          chi2         p_chi  dof  cramer_assoc    H0   p_corrected  H0_chi  \\\n",
       "59   52.780564  3.729738e-13    1      0.493182  True  1.711649e-10    True   \n",
       "61   40.077148  2.441282e-10    1      0.429753  True  3.366731e-08    True   \n",
       "63   19.370621  1.076505e-05    1      0.298773  True  3.556317e-04    True   \n",
       "64   13.705628  2.138127e-04    1      0.251316  True  4.680751e-03    True   \n",
       "65   12.515172  4.036606e-04    1      0.240153  True  5.925278e-03    True   \n",
       "55   83.097630  7.809926e-20    1      0.618820  True  2.516311e-07    True   \n",
       "51   64.841590  8.116776e-16    1      0.546634  True  8.945953e-07    True   \n",
       "53   56.005356  7.227386e-14    1      0.508025  True  2.196380e-05    True   \n",
       "48   40.858096  1.636937e-10    1      0.433920  True  6.425147e-05    True   \n",
       "52   20.555082  5.793977e-06    1      0.307772  True  5.265273e-03    True   \n",
       "25  161.686247  4.844348e-37    1      0.863191  True  1.899112e-13    True   \n",
       "26  105.414692  9.907259e-25    1      0.696981  True  5.806871e-08    True   \n",
       "33  178.147690  1.229890e-40    1      0.906067  True  5.342672e-19    True   \n",
       "31  154.954710  1.432488e-35    1      0.845031  True  4.709765e-17    True   \n",
       "32   96.699079  8.070680e-23    1      0.667546  True  1.899112e-13    True   \n",
       "37   79.968519  3.804229e-19    1      0.607057  True  5.897773e-08    True   \n",
       "38   68.219589  1.462655e-16    1      0.560692  True  7.777688e-07    True   \n",
       "40   68.219589  1.462655e-16    1      0.560692  True  7.777688e-07    True   \n",
       "35   24.151515  8.904608e-07    1      0.333613  True  5.876122e-04    True   \n",
       "18   22.094418  2.595634e-06    1      0.319089  True  1.966302e-02    True   \n",
       "1    99.350479  2.115466e-23    1      0.676636  True  8.260254e-23    True   \n",
       "3    13.017899  3.085278e-04    1      0.244929  True  1.686956e-03    True   \n",
       "6     9.582604  1.964296e-03    1      0.210142  True  9.994710e-03    True   \n",
       "2     7.947836  4.814492e-03    1      0.191379  True  2.427882e-02    True   \n",
       "\n",
       "    p_chi_corrected  \n",
       "59     1.950940e-12  \n",
       "61     1.106715e-09  \n",
       "63     3.852753e-05  \n",
       "64     6.321420e-04  \n",
       "65     1.055728e-03  \n",
       "55     7.586785e-19  \n",
       "51     5.017643e-15  \n",
       "53     4.095519e-13  \n",
       "48     7.950836e-10  \n",
       "52     2.188836e-05  \n",
       "25     1.647078e-35  \n",
       "26     1.684234e-23  \n",
       "33     8.363249e-39  \n",
       "31     3.246974e-34  \n",
       "32     9.146770e-22  \n",
       "37     3.233595e-18  \n",
       "38     9.946051e-16  \n",
       "40     9.946051e-16  \n",
       "35     3.784458e-06  \n",
       "18     1.038253e-05  \n",
       "1      2.877034e-22  \n",
       "3      8.391957e-04  \n",
       "6      4.308779e-03  \n",
       "2      1.023080e-02  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fet[(fet['H0']) & (fet['H0_chi'])].sort_values(by=['pheno', 'cramer_assoc'], ascending=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Nitrofurantoin resistance determinants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['qaccver', 'saccver', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore', 'sallseqid', 'score', 'nident', 'positive', 'gaps', 'ppos', 'qframe', 'sframe', 'qseq', 'sseq', 'qlen', 'slen', 'salltitles']\n",
    "blastp = pd.read_csv(r'C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\manuscripts\\UPEC Population Study\\code\\NIT_blastp_search.tabular', delimiter='\\t', names=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "NIT_seqs = []\n",
    "directory = r\"C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\manuscripts\\UPEC Population Study\\code\\NITREc\\Seq\\Prot\"\n",
    "for f in listdir(directory):\n",
    "    if f.endswith('.faa'):\n",
    "        with open(rf\"C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\manuscripts\\UPEC Population Study\\code\\NITREc\\Seq\\Prot\\{f}\") as handle:\n",
    "            for record in SeqIO.parse(handle, 'fasta'):\n",
    "                NIT_seqs.append({\n",
    "                    \"gene\" : f.replace('.faa', ''),\n",
    "                    \"description\" : record.description,\n",
    "                    \"seq\" : str(record.seq)\n",
    "                })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_gene_info(seq):\n",
    "    for dic in NIT_seqs:\n",
    "        if seq == dic['seq']:\n",
    "            return dic['gene'], dic['description']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\carterc\\AppData\\Local\\Programs\\Python\\Python38-32\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3062: DtypeWarning: Columns (16,17,18,22,23,25,26,27,28,30,32,35,36,37,38,39,40,42,43,45,46,47,51,54,56,57,58,61,62,63,65,66,68,69,70,74,75,76,77,78,80,82,83,85,87,88,89,90,91,92,94,95,96,98,100,101,102,103,104,107,109,112,116,118,120,130,132,137,140,144,150,156,158,161,164,166,167,169,172,173,176,182,184,185,186,187,189,191,193,196,208,210,215,223,224,228,230,231) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    }
   ],
   "source": [
    "gpa = pd.read_csv(r'C:\\Users\\carterc\\OneDrive - Norwich BioScience Institutes\\Data\\UPEC population study\\dataset\\gene_presence_absence.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "ID_key_dict = gpa.filter(like='GL').iloc[0].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_value(value):\n",
    "    return re.search('[A-Z]*', value).group(0)\n",
    "\n",
    "map_rename = lambda x : mapped_IDs[rename_value(x)]\n",
    "mapped_IDs = {rename_value(v) : k for k,v in ID_key_dict.items()}\n",
    "blastp = blastp.drop(blastp[blastp['saccver'].str.startswith('HNPKFMLJ')].index)\n",
    "blastp[\"isolate\"] = blastp['saccver'].map(map_rename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "blastp[['gene', 'description']] = pd.DataFrame(blastp['qseq'].map(fetch_gene_info).to_list(), index=blastp.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "blastp_subset = blastp[(blastp['ppos'] == 100) & (blastp['pident'] == 100) & (blastp['qlen'] == blastp['slen'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      isolate                                        description     gene\n",
      "1342    GL116  NfsA nfsA|IN09|2|287989-288480|-|ICL2020_00865...  NfsA__R\n",
      "5150     GL14  NfsA nfsA|EC958|NZ_HG941718.1|900522-901244|+|...  NfsA__S\n",
      "10266     GL4  NfsB nfsB|IN06|1|705493-705774|+|ICL2020_00712...  NfsB__R\n",
      "10619   GL160  NfsB nfsB|IN07|2|122188-122619|+|ICL2020_00466...  NfsB__R\n",
      "15019    GL14  NfsB nfsB|ATCC25922|NZ_CP009072.1|4640270-4640...  NfsB__S\n",
      "18477    GL14  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18479    GL13  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18484    GL12  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18488     GL3  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18489    GL11  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18494     GL2  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18498     GL5  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18501     GL4  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18539   GL160  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18602   GL116  RibE ribE|ATCC25922|NZ_CP009072.1|4779388-4779...  RibE__S\n",
      "18689     GL1  RibE ribE|UTI89|NC_007946.1|448891-449361|+|UT...  RibE__S\n"
     ]
    }
   ],
   "source": [
    "with pd.option_context('display.max_rows', None, 'display.max_columns', None):\n",
    "    print(blastp_subset[blastp_subset['isolate'].isin(amr[amr['NIT']=='R'].index)][['isolate', 'description', 'gene']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "86ac97f5aa0aa7efa4e154d9a1a31a60e75f86ffa61201e677a1a795af44ce38"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
